### ISSUE TITLE
Migration of Java Code to Kotlin for Enhanced Code Consistency and Maintainability

### ISSUE THREAD

**Issue by Saifuddin53** [commons-app/apps-android-commons#5928]
### What is the user problem or growth opportunity you want to see solved?

The current codebase consists of a mix of Java and Kotlin, which introduces inconsistencies and can make the code harder to maintain and extend. Migrating fully to Kotlin would streamline the code, reduce verbosity, and provide a more cohesive developer experience. This transition would also open up future opportunities to leverage Kotlin Multiplatform (KMP), facilitating easier expansion to iOS if desired.

### How do you know that this problem exists today? Why is this important?

Android development is increasingly Kotlin-centric, as it’s the recommended language with strong support for modern Android features, readability, and efficiency improvements. Having a consistent codebase fully in Kotlin will improve maintainability and simplify onboarding. 

### Who will benefit from it?

- **Developers:** A fully Kotlin-based codebase will improve readability, maintainability, and efficiency, enhancing the developer experience.
- **Project Maintainability:** This alignment with modern Android standards will simplify contributions and align with future technology trends.
- **End Users:** Indirectly, users will benefit from a well-maintained app that can more easily adopt new features and improvements.

### Anything else you would like to add?

Once the migration to Kotlin is complete, further modernization could include adopting Jetpack Compose for the UI layer, reducing XML dependency, and simplifying the UI structure. This approach would prepare the project for KMP, creating a strong foundation for potential iOS app development in the future.

**Comment by misaochan** [comment]
I agree wholeheartedly with your points, and we started the process several years ago. It will take a long time and quite a lot of work, however, as it involves rewriting large portions of the app. Would you be interested in working on this, @Saifuddin53 ?

**Comment by Saifuddin53** [comment]
Yes @misaochan @nicolas-raoul, I'm interested in contributing to this migration process. 

I plan to start with the utility/helper classes, as they are more isolated and easier to migrate. After completing that, I’ll adopt a module-wise approach, focusing on one module at a time to ensure smooth transitions and minimal disruptions.

Alternatively, if needed, I can prioritize specific areas, such as the data layer or UI components. Since this involves a large amount of work, anyone interested in joining the migration process is more than welcome to collaborate. It would be great to share the workload and move faster while ensuring high-quality results.

**Comment by misaochan** [comment]
That sounds great, thank you very much. :) I'll assign you to this @Saifuddin53 , but if anyone else is interested, please feel free to chime in. Unlike other issues, I think it should be fine for multiple people to work on this one, as it is a huge task. We can probably add a checklist later to track the PRs as they come in.

**Comment by Saifuddin53** [comment]
Thank you so much for assigning me to this, @misaochan 

Also, for anyone who’d like to join this migration effort:

1. **Check if someone is already working on the file/module** you’re interested in. This can be done by looking at the existing issues or PRs related to the migration.  

2. **Create a ticket/issue for the files or modules you’re taking on, if no one is already working on them.** In the ticket:  
   - Mention the parent migration ticket.  
   - Specify the files/modules you plan to work on.  

This approach will help us stay coordinated and avoid overlaps.

**Comment by neeldoshii** [comment]
Working on  Bookmark module & Achievement Module

**Comment by neeldoshii** [comment]
I guess its better to migrate a class by class to java to kotlin rather than a module by module. isn't it? Focusing more on refactoring our previous java code too as things in kotlin gives us lot of advantage to null pointer and fixing our lint issues since its a whole rewrite. What's your take?

**Comment by Saifuddin53** [comment]
> Alternatively, if needed, I can prioritize specific areas, such as the data layer or UI components.

Of course, that's a valid point, @neeldoshii. You can also take that approach too, as mentioned earlier. A hybrid approach might work well here. However, there are also benefits to the module-by-module approach, it allows us to complete logical chunks of the application which can be tested and reviewed in isolation.

**Comment by nicolas-raoul** [comment]
I think both approaches are valid. :-)
Of course, refactoring to take advantages of all Kotlin features is the best (either at class level or module level).
Taking advantages of _all_ Kotlin features requires deep Kotlin expertise, so I think perfection is not _required_ here, hopefully further improvements can be made even after migration.

To anyone: Would you mind writing down here the Java->Kotlin migration best practices that are specific to our app? For instance yesterday's great `_binding = null` tip. Or maybe link to an external document if you find one that is relevant enough.

Thanks all! :-)

**Comment by chrisdebian** [comment]
@Saifuddin53 Hi, just trying to help prioritize what development work gets done next. Would you be able to have a look at this [link](https://github.com/commons-app/apps-android-commons/issues/6034) and potentially add your thoughts/ priorities to the spreadsheet?

Thanks,

Chris.

**Comment by chrisdebian** [comment]
I have a broad awareness of this issue, but not technical knowledge. If the project has agreed that moving to Kotlin is a sensible idea (it seems to be), then what do we need to do to make this happen?

Thanks,

Chris.

**Comment by psh** [comment]
> If the project has agreed that moving to Kotlin is a sensible idea (it seems to be), then what do we need to do to make this happen?

Chris, read back up this issue, and see all of the merges and people posting to claim packages?  It already *is* happening; the grass-roots / self-organizing movement is doing great.  Make a cup of tea, sit back and maybe have some fun - visualize the updates to the project with a tool like [Gource](https://gource.io/) 😄

**Comment by chrisdebian** [comment]
Thanks, Paul; i've already moved on. The 500 bugs will sort themselves, even the ones that are 10 years old. I'll continue as a Commons user, and leave organising to those best able to do it.

Chris

**Comment by psh** [comment]
Quick check on status - searching for a count of Java files in packages - 

| Package                        | java files |
| ------------------------------ | ---------- |
| bookmarks	                     | 3          |
| bookmarks/items	             | 4          |
| bookmarks/locations	         | 4          |
| bookmarks/pictures	         | 4          |
| concurrency	                 | 4          |
| contributions	                 | 16         |
| explore	                     | 5          |
| explore/depictions	         | 1          |
| explore/map	                 | 5          |
| explore/recentsearches	     | 3          |
| media	                         | 7          |
| media/zoomControllers/gestures | 2          |
| media/zoomControllers/zoomable | 9          |
| nearby	                     | 15         |
| nearby/contract	             | 1          |
| nearby/fragments	             | 1          |
| profile	                     | 1          |

Great work all!  Let's keep it going!

**Comment by Saifuddin53** [comment]
@psh bookmarks items, locations, pictures would also be covered in #6017 :)

**Comment by Sujal-Gupta-SG** [comment]
@nicolas-raoul Can i also work on this issue so that work get completed fastly

**Comment by Saifuddin53** [comment]
Yes @Sujal-Gupta-SG, feel free to go ahead and start working on it. Choose a package or file you'd like to work on, create an issue for it, and begin your work. Once you're done, submit a pull request. 🙂

**Comment by Sujal-Gupta-SG** [comment]
@psh I have done profile package, can you assign the issue #6118 and check whether it satisfy the condition i have raised the pull request also #6119, if not then i can again work on it

**Comment by Sujal-Gupta-SG** [comment]
I am migrating contribution folder if anyone is doing already let me know

**Comment by psh** [comment]
Based on what is currently on the `main` branch, I am seeing the remaining Java files needing conversion, based on running the Unix command 

```bash
find . -name \*.java >> remaining.txt 
```

### Nearby
```
nearby/NearbyController.java
nearby/PlacesLocalDataSource.java
nearby/contract/NearbyParentFragmentContract.java
nearby/Sitelinks.java
nearby/Label.java
nearby/Place.java
nearby/NearbyfilterSearchListView.java
nearby/PlaceDao.java
nearby/PageEditHelper.java
nearby/CheckBoxTriStates.java
nearby/NearbyNotificationCardView.java
nearby/PlacesRepository.java
nearby/NearbyPlaces.java
nearby/MarkerPlaceGroup.java
nearby/NearbyFilterSearchRecyclerViewAdapter.java
nearby/NearbyFilterState.java
```

### Explore
~explore/ExploreListRootFragment.java
explore/ExploreFragment.java
explore/depictions/WikidataItemDetailsActivity.java
explore/ParentViewPager.java
explore/map/ExploreMapContract.java
explore/map/ExploreMapFragment.java
explore/map/ExploreMapPresenter.java
explore/map/ExploreMapCalls.java
explore/map/ExploreMapController.java
explore/ExploreMapRootFragment.java
explore/recentsearches/RecentSearchesDao.java
explore/recentsearches/RecentSearchesContentProvider.java
explore/recentsearches/RecentSearchesFragment.java
explore/SearchActivity.java~

### Bookmarks
~bookmarks/pictures/BookmarkPicturesContentProvider.java
bookmarks/pictures/BookmarkPicturesDao.java
bookmarks/pictures/BookmarkPicturesController.java
bookmarks/pictures/BookmarkPicturesFragment.java
bookmarks/BookmarkFragment.java
bookmarks/BookmarkListRootFragment.java
bookmarks/items/BookmarkItemsController.java
bookmarks/items/BookmarkItemsContentProvider.java
bookmarks/items/BookmarkItemsDao.java
bookmarks/items/BookmarkItemsFragment.java
bookmarks/BookmarksPagerAdapter.java~

### Media

~media/CaptionListViewAdapter.java
media/MwParseResult.java
media/CustomOkHttpNetworkFetcher.java
media/Caption.java
media/MwParseResponse.java
media/CommonsWikibaseItem.java
media/MediaDetailPagerFragment.java~

### Top Level

~WelcomeActivity.java
MapController.java
License.java
ViewHolder.java
Utils.java
MvpView.java
BasePresenter.java
WelcomePagerAdapter.java
OkHttpConnectionFactory.java
ViewPagerAdapter.java~

**Comment by psh** [comment]
> Sonal is currently working on Nearby, that's the topic of this summer's Oureachy.
> 
> How about converting converting another area, then Nearby after Outreachy which ends in August? 🙂 

 _Originally posted by @nicolas-raoul in [#6391](https://github.com/commons-app/apps-android-commons/issues/6391#issuecomment-3152636514)_

Is now a good time to look at completing the migration?  We're *so* close!

**Comment by nicolas-raoul** [comment]
Now is Hacktober so we receive more pull requests than usual, but then GSoC candidates will start coming, so I would say now is a good time. :-)

**Comment by RitikaPahwa4444** [comment]
Just wanted to thank everyone for contributing to this migration; I had never expected it to be this quick! It's so good to see that our codebase is mostly in Kotlin now ❤️

**Comment by neeldoshii** [comment]
> Now is Hacktober so we receive more pull requests than usual, but then GSoC candidates will start coming, so I would say now is a good time. :-)

I won’t be able to contribute much this time due to a busy schedule, but I can timebox a few hours and would love to help as a mentor, maintainer, or with code reviews. I just want to thank everyone who has contributed to this migration.  

One year ago, the repository was 25% Java legacy, two years ago it was almost 80%. Now it’s only 4%! I never expected this to get wrapped up so quickly. Kudos to everyone! 🎉
